In [1]:
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()
X, y = data.data, data.target

print("Samples:", X.shape[0])
print("Features:", X.shape[1])
print("Target classes:", list(data.target_names))  # ['malignant', 'benign']
print("Feature names:", list(data.feature_names))  # 30 names as above


Samples: 569
Features: 30
Target classes: [np.str_('malignant'), np.str_('benign')]
Feature names: [np.str_('mean radius'), np.str_('mean texture'), np.str_('mean perimeter'), np.str_('mean area'), np.str_('mean smoothness'), np.str_('mean compactness'), np.str_('mean concavity'), np.str_('mean concave points'), np.str_('mean symmetry'), np.str_('mean fractal dimension'), np.str_('radius error'), np.str_('texture error'), np.str_('perimeter error'), np.str_('area error'), np.str_('smoothness error'), np.str_('compactness error'), np.str_('concavity error'), np.str_('concave points error'), np.str_('symmetry error'), np.str_('fractal dimension error'), np.str_('worst radius'), np.str_('worst texture'), np.str_('worst perimeter'), np.str_('worst area'), np.str_('worst smoothness'), np.str_('worst compactness'), np.str_('worst concavity'), np.str_('worst concave points'), np.str_('worst symmetry'), np.str_('worst fractal dimension')]


In [2]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load dataset
data = load_breast_cancer()

# Create DataFrame with features
df = pd.DataFrame(data.data, columns=data.feature_names)

# Add target column (0 = malignant, 1 = benign)
df['target'] = data.target

# Display the first 5 rows
print(df.head())


   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0             

# Task: Load the Breast Cancer dataset and create a basic logistic regression model without preprocessing.

In [3]:
# Import necessary libraries
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
import pandas as pd

# Load the dataset
data = load_breast_cancer()
X = data.data  # Features
y = data.target  # Target (0 = malignant, 1 = benign)
print(data)

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train logistic regression model
model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)
print(model.predict_proba(X_test[:5]))


# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print("Confusion Matrix:")
print(cm)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



### Explanation:
- load_breast_cancer() loads a binary classification dataset
- X contains 30 features about tumor characteristics
- y contains binary labels (0 = malignant, 1 = benign)
- LogisticRegression is used for binary classification
- confusion_matrix shows:

    [True Negative, False Positive]

    [False Negative, True Positive]



## Notes:
1- Logistic Regression is for classification, Linear Regression for regression

2- Outputs probabilities between 0 and 1 using sigmoid function

3- Perfect for binary classification problems

4- Interpretable and fast for small to medium datasets

5- Requires feature scaling for best performance

6- Can be extended to multi-class problems

# Task: Calculate and interpret precision, recall, and F1-score from the confusion matrix.

In [4]:
from sklearn.metrics import classification_report

# Calculate evaluation metrics
tn, fp, fn, tp = cm.ravel()

precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1 = 2 * (precision * recall) / (precision + recall)

print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

# Using sklearn's built-in function
print("\nDetailed Classification Report:")
print(classification_report(y_test, y_pred, target_names=['Malignant', 'Benign']))

Precision: 0.9459
Recall: 0.9859
F1-Score: 0.9655

Detailed Classification Report:
              precision    recall  f1-score   support

   Malignant       0.97      0.91      0.94        43
      Benign       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114




- Precision: How many predicted positives are actually positive
- Recall: How many actual positives are correctly predicted
- F1-Score: Harmonic mean of precision and recall
- Macro Average: Simple average of both classes (0.97 + 0.97) / 2 = 0.97

- Weighted Average: Average weighted by support (43×0.97 + 71×0.97) / 114 = 0.97

Interpretation:
- High precision: Few false alarms
- High recall: Misses few positive cases
- Good model should have both high precision and recall


Key Takeaways


1.   Precision: Quality of positive predictions → "How trustworthy are our 'yes' answers?"

2. Recall: Coverage of actual positives → "How many actual 'yes' cases did we find?"

3. F1-Score: Balanced measure when both matter equally

4. Context Matters: In healthcare, recall is often more important than precision

5. Trade-offs: Improving one metric often worsens another


For the breast cancer model(medical in general):

  1. High recall means you're good at detecting actual cancer cases (safety first!)

  2. High precision means you're not causing unnecessary worry with false alarms

  3. Your balanced F1-score suggests good overall performance for medical use

# Task: Apply logistic regression to classify three types of iris flowers.

In [5]:
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
import pandas as pd
# Load iris dataset
iris = load_iris()
X_iris = iris.data
y_iris = iris.target

# Split the data
X_train_iris, X_test_iris, y_train_iris, y_test_iris = train_test_split(
    X_iris, y_iris, test_size=0.3, random_state=42
)

# Create multi-class logistic regression model
# multi_class='multinomial' enables Softmax for multi-class classification
model_iris = LogisticRegression(multi_class='multinomial', random_state=42)
model_iris.fit(X_train_iris, y_train_iris)

# Predictions
y_pred_iris = model_iris.predict(X_test_iris)

# Evaluate
accuracy_iris = accuracy_score(y_test_iris, y_pred_iris)
cm_iris = confusion_matrix(y_test_iris, y_pred_iris)

print(f"Iris Dataset Accuracy: {accuracy_iris:.4f}")
print("Iris Confusion Matrix:")
print(cm_iris)
print(f"Class names: {iris.target_names}")

Iris Dataset Accuracy: 1.0000
Iris Confusion Matrix:
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]
Class names: ['setosa' 'versicolor' 'virginica']


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(



##Multi-class Classification:

- Softmax function generalizes sigmoid for multiple classes
- multi_class='multinomial' uses Softmax regression
- Model outputs probabilities for each class
- Final prediction is class with highest probability

###The confusion matrix for 3 classes shows:
- Diagonal: Correct predictions for each class
- Off-diagonal: Misclassifications between classes


# Task: Demonstrate the importance of feature scaling for logistic regression.

In [6]:
from sklearn.preprocessing import StandardScaler
import numpy as np

# Create sample data with different scales
np.random.seed(42)
X_uneven = np.column_stack([
    np.random.normal(100, 10, 100),  # Large scale (mean=100)
    np.random.normal(0, 1, 100)      # Small scale (mean=0)
])
y_uneven = (X_uneven[:, 0] + X_uneven[:, 1] > 100).astype(int)

# Split data
X_train_u, X_test_u, y_train_u, y_test_u = train_test_split(
    X_uneven, y_uneven, test_size=0.3, random_state=42
)

# Model without scaling
model_no_scale = LogisticRegression(random_state=42)
model_no_scale.fit(X_train_u, y_train_u)
accuracy_no_scale = accuracy_score(y_test_u, model_no_scale.predict(X_test_u))

# Model with scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_u)
X_test_scaled = scaler.transform(X_test_u)

model_scaled = LogisticRegression(random_state=42)
model_scaled.fit(X_train_scaled, y_train_u)
accuracy_scaled = accuracy_score(y_test_u, model_scaled.predict(X_test_scaled))

print(f"Accuracy without scaling: {accuracy_no_scale:.4f}")
print(f"Accuracy with scaling: {accuracy_scaled:.4f}")
print(f"Improvement: {accuracy_scaled - accuracy_no_scale:.4f}")

Accuracy without scaling: 0.9667
Accuracy with scaling: 1.0000
Improvement: 0.0333


## Why Scaling Matters:
1. Logistic regression uses gradient descent
2. Features with larger scales dominate the optimization
3. Scaling ensures all features contribute equally
4. StandardScaler transforms data to mean=0, std=1

StandardScaler formula:
    z = (x - mean) / std

Always:

- fit_transform on training data
- transform on test data (using training parameters)


In [7]:

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

data = load_breast_cancer()
X = data.data
y = data.target

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("logreg", LogisticRegression(max_iter=1000, random_state=42))
])

pipeline.fit(X_train, y_train)

y_pred = pipeline.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print(f"Accuracy: {accuracy:.4f}")
print("Confusion Matrix:")
print(cm)
print("\nClassification report:")
print(classification_report(y_test, y_pred, target_names=data.target_names))


Accuracy: 0.9737
Confusion Matrix:
[[41  2]
 [ 1 70]]

Classification report:
              precision    recall  f1-score   support

   malignant       0.98      0.95      0.96        43
      benign       0.97      0.99      0.98        71

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



In [8]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
import numpy as np
import pandas as pd


data = load_breast_cancer()
X = data.data
y = data.target
feature_names = data.feature_names


pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("model", LogisticRegression(max_iter=1000, random_state=42))
])


pipeline.fit(X, y)


model = pipeline.named_steps["model"]
coefficients = model.coef_[0]


importance = pd.DataFrame({
    "Feature": feature_names,
    "Coefficient": coefficients,
    "Absolute Importance": np.abs(coefficients)
})


importance_sorted = importance.sort_values(by="Absolute Importance", ascending=False)

print(importance_sorted)


                    Feature  Coefficient  Absolute Importance
21            worst texture    -1.320556             1.320556
10             radius error    -1.289261             1.289261
20             worst radius    -1.026586             1.026586
13               area error    -0.998861             0.998861
23               worst area    -0.994671             0.994671
7       mean concave points    -0.962849             0.962849
27     worst concave points    -0.925175             0.925175
28           worst symmetry    -0.888711             0.888711
26          worst concavity    -0.880223             0.880223
6            mean concavity    -0.855295             0.855295
22          worst perimeter    -0.820397             0.820397
15        compactness error     0.744077             0.744077
19  fractal dimension error     0.681746             0.681746
12          perimeter error    -0.671895             0.671895
24         worst smoothness    -0.665715             0.665715
5       